### DISCUSSION TITLE
Request: PlantUMLのアクティビティ図にて`:`を利用すると絵文字の`:`として認識されて絵文字の入力補完される挙動を改善する

### DISCUSSION THREAD

**Discussion by Akira-Demizu** [growilabs/growi#7729] [Ideas / Requests] [answered]
Environment
------------

### Host
| item     | version |
| ---      | --- |
|OS        |Amazon Linux 2|
|GROWI     |5.1.4|
|node.js   |16.17.0|
|npm       |8.15.0|
|yarn      |1.22.19|
|Using Docker|yes|
|Using [growi-docker-compose][growi-docker-compose]|yes|

[growi-docker-compose]: https://github.com/weseek/growi-docker-compose

*(Accessing https://{GROWI_HOST}/admin helps you to fill in above versions)*

### Client

| item     | version |
| ---      | --- |
|OS        |macOS Monterey バージョン12.6|
|browser   |Google Chrome 106.0.5249.119|

How to reproduce? (再現手順)
---------------------------

1. 任意のページにてPlantUMLのアクティビティ図を利用します。
    ```bash
    @startuml
    :tes
    @enduml
    ```

What happens? (症状)
---------------------
`:`を利用すると絵文字の`:`として認識されて絵文字の入力補完される。

![スクリーンショット 2022-10-26 15 58 32](https://user-images.githubusercontent.com/38356118/197959708-a4b8a30a-912a-4150-9a77-461b668fb550.png)

What is the expected result? (期待される動作)
-------------------------------------------
Plantumlコード内では絵文字を利用することはなさそうなので、`@startuml` ~ `@enduml`の間では絵文字が反応しない。

**Comment by yuki-takei** [comment]
ご意見ありがとうございます。こちらはバグではないのでタイトルとラベルを変更させていただきました。

現状この挙動への対応は難しいです。エディタである CodeMirror 的に「今何のブロックを編集しているのか」という情報をつぶさに取る機能がないためです。(table syntax の中にカーソルがあるとか drawio syntax の中にカーソルがある、といった判定をする実装は今ありますが、実は結構無理をしています)

代案としては、emoji の suggest の挙動を変えるという案は開発チーム内で出ています。

**現状**: emoji picker のポップアップが出る(エディタからフォーカスが外れる)
**代案**: 以下のスクリーンショットのような CodeMirror の hint を使う実装を追加する(GROWI の過去のバージョンはこれでした)

![image](https://user-images.githubusercontent.com/1638767/198001194-4b0a6c27-5f05-4e8c-8b89-7e48a9fc60c0.png)

代案でも PlantUML の図を編集している最中に suggest が出るのは回避できませんが、hint が出ても続けてエディタにフォーカスが当たっているのでストレスが少なくなるのではないかと予想しています。

**Comment by Akira-Demizu** [comment]
@yuki-takei 
ご確認ありがとうございます。
また、改善案をご検討いただきありがとうございます。

私からのご提案の代案としては、そもそもemojiを利用するかどうかを管理者が設定(選択)できるとありがたいです。
以前まではそもそもemojiがない状態で問題なく利用できていたので、昔から利用している方にも馴染みやすいと考えております。
ご検討のほどお願いいたします。

**Comment by yuki-takei** [comment, marked as answer]
2023Q4 リリース予定の GROWI v7 で、CodeMirror の autocompletion 機能を利用した実装になる予定です。
